File: alg_ds_1/linked_list/linked_list.go
package main

import (
	"errors"
)

type Node struct {
	next  *Node
	value int
}

type LinkedList struct {
	head *Node
	tail *Node
}

func (l1 *LinkedList) AddInTail(item *Node) {
	item.next = nil
	if l1.head == nil {
		l1.head = item
		l1.tail = item
	} else {
		l1.tail.next = item
		l1.tail = item
	}
}

func (l1 *LinkedList) Count() int {
	count := 0
	cur := l1.head
	for cur != nil {
		count++
		cur = cur.next
	}
	return count
}

func (l1 *LinkedList) Find(n int) (*Node, error) {
	curr := l1.head
	for curr != nil {
		if curr.value == n {
			return curr, nil
		}
		curr = curr.next
	}
	return nil, errors.New("node not found")
}

func (l1 *LinkedList) FindAll(n int) []*Node {
	var nodes []*Node
	curr := l1.head
	for curr != nil {
		if curr.value == n {
			nodes = append(nodes, curr)
		}
		curr = curr.next
	}
	return nodes
}

func (l1 *LinkedList) Delete(n int, all bool) {
	if l1.head == nil {
		return
	}
	// Удаление из начала списка
	for l1.head != nil && l1.head.value == n {
		l1.head = l1.head.next
		if l1.head == nil {
			l1.tail = nil
			return
		}
		if !all {
			if l1.head == nil {
				l1.tail = nil
			}
			return
		}
	}
	// Удаление из "середины"
	prev := l1.head
	if prev == nil {
		l1.tail = nil
		return
	}
	curr := prev.next
	for curr != nil {
		if curr.value == n {
			prev.next = curr.next
			if curr == l1.tail {
				l1.tail = prev
			}
			if !all {
				return
			}
			curr = prev.next
		} else {
			prev = curr
			curr = prev.next
		}
	}
}

func (l1 *LinkedList) Insert(after *Node, add *Node) {
	if l1.head == nil {
		l1.head = add
		l1.tail = add
		add.next = nil
		return
	}
	add.next = after.next
	after.next = add
	if l1.tail == after {
		l1.tail = add
	}
}

func (l1 *LinkedList) InsertFirst(first *Node) {
	first.next = l1.head
	l1.head = first
	if l1.tail == nil {
		l1.tail = first
	}
}

func (l1 *LinkedList) Clean() {
	l1.head = nil
	l1.tail = nil
}

File: alg_ds_1/python_version/01_linked_list/linked_list.py
class Node:

    def __init__(self, v):
        self.value = v
        self.next = None

class LinkedList:

    def __init__(self):
        self.head = None
        self.tail = None

    def add_in_tail(self, item):
        if self.head is None:
            self.head = item
        else:
            self.tail.next = item
        self.tail = item

    def print_all_nodes(self):
        node = self.head
        while node != None:
            print(node.value)
            node = node.next

    def find(self, val):
        node = self.head
        while node is not None:
            if node.value == val:
                return node
            node = node.next
        return None

    # Задание на курсе: 1
    # Задача: 1.4
    # Название: поиск всех узлов по значению
    # Временная сложность: O(n)
    # Пространственная сложность: O(n)
    def find_all(self, val):
        res = []
        if self.head is None:
            return res
        cur = self.head
        while cur:
            if cur.value == val:
                res.append(cur)
            cur = cur.next
        return res

    # Задание на курсе: 1
    # Задачи: 1.1, 1.2
    # Название: удаление первого или всех узлов по значению
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def delete(self, val, all=False):
        while self.head is not None and self.head.value == val:
            self.head = self.head.next
            if not all:
                if self.head is None:
                    self.tail = None
                return

        if self.head is None:
            self.tail = None
            return

        cur = self.head

        while cur.next is not None:
            if cur.next.value == val:
                if cur.next is self.tail:
                    self.tail = cur
                cur.next = cur.next.next

                if not all:
                    return
            else:
                cur = cur.next

    # Задание на курсе: 1
    # Задача: 1.3
    # Название: очистка списка
    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def clean(self):
        self.head = None
        self.tail = None

    # Задание на курсе: 1
    # Задача: 1.5
    # Название: вычисление длины списка
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def len(self):
        count = 0
        cur = self.head
        while cur is not None:
            count += 1
            cur = cur.next
        return count

    # Задание на курсе: 1
    # Задача: 1.6
    # Название: вставка узла
    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def insert(self, afterNode, newNode):
        if afterNode is None:
            newNode.next = self.head
            self.head = newNode
        else:
            newNode.next = afterNode.next
            afterNode.next = newNode

        if newNode.next is None:
            self.tail = newNode
